### ISSUE TITLE
[Question] Proper data format for GRPO Agent Training

### ISSUE THREAD

**Issue by August-murr** [huggingface/trl#2809]
At first I wanted to use smolagents Agents to train with GRPO but then I realized the enormous amounts of prompts used to guide the Agent aren't useful for training an agent and will complicate everything since the Agent learns to do everything on it's own. 

so instead I want to proceed with a simple format like the 
```json
[
  {
    "role": "user",
    "content": "Can you help me calculate the factorial of 5 ?"
  },
  {
    "role": "assistant",
    "content": "Sure, I'll write a recursive function for that."
  },
  {
    "role": "assistant",
    "content": "Here's the code:"
  },
  {
    "role": "tool",
    "content": "```python\n# Recursive factorial function\n\ndef factorial(n):\n    if n <= 1:\n        return 1\n    return n * factorial(n-1)\n\nprint(factorial(5))\n```"
  },
  {
    "role": "tool",
    "content": "Output: 120"
  },
  {
    "role": "assistant",
    "content": "The code runs correctly and returns 120. Would you like me to explain how it works?"
  }
]
```
and this raises some questions:

1. Would this multi-turn format (including multiple tool messages—one for code and one for its output) work with the current GRPO implementation?

2. Is there a recommended or “correct” format for such agentic data that includes tool usage for GRPO?
 
3. Are there any special considerations for the tool responses? For instance, should we exclude those tokens from KL divergence computations (as @xiangjjj mentioned in #2723 ) or handle them differently with attention masks?

**Comment by qgallouedec** [comment]
> 1. Would this multi-turn format (including multiple tool messages—one for code and one for its output) work with the current GRPO implementation?

I think it _should_ work.

> 2. Is there a recommended or “correct” format for such agentic data that includes tool usage for GRPO?

Not that I am aware of. As long as the format is supported by the chat template of the model.

> 3. Are there any special considerations for the tool responses? For instance, should we exclude those tokens from KL divergence computations (as [@xiangjjj](https://github.com/xiangjjj) mentioned in [Training Agents with GRPO #2723](https://github.com/huggingface/trl/issues/2723) ) or handle them differently with attention masks?

My hunch is that, indeed, these tokens should be masked for the loss calculation (not just the KL part, also the advantage part). But that it's probably OK not to mask them. My feeling is that even if these tokens won't be generated by the model, it may still help to train the model to predict them, or at least it won't hurt the learning process.

**Comment by qgallouedec** [comment]
Hey @August-murr I've managed to make an agentic GRPO with just 2 lines change in GRPO; check this!
It's a toy example where the model is trained to find the max value of an unknown function but it's funny, because it works!

```python
from datasets import Dataset
from trl import GRPOConfig, GRPOTrainer
import re
import json


def get_value(x: float) -> float:
    """
    Get the value of the function at x.

    Args:
        x: The input value.

    Returns:
        The value of the function at x.
    """
    return max(-((x / 5) ** 2) + x + 1, 0.0)


def agent_reward(completions, **kwargs):
    rewards = []
    for completion in completions:
        content = completion[0]["content"]
        # Regex pattern to find the JSON inside <tool_call>...</tool_call>
        match = re.search(r"<tool_call>\s*(\{.*?\})\s*</tool_call>", content, re.DOTALL)
        if not match:
            rewards.append(-100)
            continue

        # Try to parse the JSON content
        try:
            # Parse the JSON content
            json_data = json.loads(match.group(1))
        except json.JSONDecodeError:
            rewards.append(-80)
            continue

        # Check if the function name is "get_value"
        function_name = json_data.get("name", "")
        if function_name != "get_value":
            rewards.append(-60)
            continue

        # Get the value of "x" argument
        value = json_data.get("arguments", {}).get("x")
        if value is None:
            rewards.append(-40)
            continue

        # Check if the value is a float
        if not isinstance(value, (int, float)):
            rewards.append(-20)
            continue

        rewards.append(get_value(float(value)))
    return rewards


dataset = Dataset.from_list(
    [{"prompt": [{"role": "user", "content": "Call the function get_value with any value."}]}] * 200
)


def main():
    training_args = GRPOConfig(
        output_dir="Qwen2.5-0.5B-GRPO-agent",
        logging_steps=5,
        gradient_accumulation_steps=4,
        max_completion_length=128,
        max_prompt_length=128,
        bf16=True,
        log_completions=True,
    )
    trainer = GRPOTrainer(
        model="Qwen/Qwen2.5-0.5B-Instruct",
        reward_funcs=agent_reward,
        args=training_args,
        train_dataset=dataset,
        tools=[get_value],
    )

    trainer.train()


if __name__ == "__main__":
    main()
```

<img width="1275" alt="Image" src="https://github.com/user-attachments/assets/ef135060-514e-457b-a2d6-97ac24aacde2" />

**Comment by NickyDark1** [comment]
I don't know if I'm wrong, but in the url:
```py
https://github.com/huggingface/trl/blob/main/trl/trainer/grpo_trainer.py
```

I notice that (GRPOTrainer, tools) 'tools=[get_value]' is not there.

**Comment by qgallouedec** [comment]
Yes:

>  with just 2 lines change in GRPO

described here https://github.com/huggingface/trl/issues/2809#issuecomment-2646369933

**Comment by NickyDark1** [comment]
Thank you very much for the prompt response.

but then 'get_value' is applied only to all rows, how would it be to be able to make it have different call functions?

**Comment by willccbb** [comment]
idea with this PR (https://github.com/huggingface/trl/pull/2810) is to offload all of the rollout logic into an `Environment` object which just needs to return `completion_ids` for each prompt. Then it's up to the user to determine how the rollouts are constructed + we don't need to hardcode any assumptions about tool use into TRL. Would this address most of what people are hoping for from multi-step/agentic use cases?

**Comment by bevren** [comment]
can we append output of the function to the context? like <output> output of the function </output> during generation?

**Comment by willccbb** [comment]
The PR is not merged yet, would appreciate any comments over there (or if someone can be assigned to review). But yes, it would allow full rollout logic to be customized by the user (with outcome rewards over the entire trajectory). Separately, I am working on a [repo](https://github.com/willccbb/verifiers) which would plug into TRL via this interface, and would provide some useful primitives for building these kinds of interaction loops.
